### DISCUSSION TITLE
Using functions for security & business logic enforcement -- safe?

### DISCUSSION THREAD

**Discussion by wrightd** [supabase/supabase#12246] [Questions]
I've started using supabase for a Jamstack-style (client only) application, but am struggling with how to enforce security & business logic using just RLS. Honestly, it gives me the heebie jeebies to allow client javascript to make direct database calls even with some security, but more importantly I struggled to make more than trivial rules work if I wanted to incorporate anything resembling business logic rather than just simple access checks. 
I've moved to using RLS+ database functions, where DB functions are used for any data modification and RLS policies never allow insert/update/delete by users. 
For example, I've included a simple function below to create a team, which enforces several checks around number of teams and team names, and does a couple inserts needed to create the team. It executes with definer security since the caller doesn't have any insert permissions.

**My question is... is this safe?** The supabase docs give some basic explanation of security using definer functions (basically, they will bypass RLS), but it seems like as long as my business logic in the function is correct, my app should be as well. That said, I want to sanity check this assumption before I build my whole app this way. Also, assuming it is correct... it seems like an approach a lot of people may want to take to get around some of the limitations of RLS when it comes to enforcing business logic for modification operations rather than just giving clients full run over (even their subset) of the database. I'm surprised it isn't called out more directly in the documentation around RLS/security.

```
create or replace function create_team(p_name text)
returns int8 
language plpgsql
security definer set search_path = public
as $$
DECLARE
v_numteams int;
v_teamID int8;
begin
  IF (auth.uid() isnull) THEN
    RAISE EXCEPTION SQLSTATE '90001' USING MESSAGE = 'No logged in user';
  END IF;
  select count(*) into v_numteams from teams where creator = auth.uid() and deleted=false;
  IF (v_numteams > 5) THEN
    RAISE EXCEPTION SQLSTATE '90001' USING MESSAGE = 'Too many teams created';
  END IF;
  IF (LENGTH(p_name) > 50) THEN
    RAISE EXCEPTION SQLSTATE '90001' USING MESSAGE = 'Team name too long';
  END IF;
  select count(*) into v_numteams from teams where teams.name=p_name and deleted=false;
  IF (v_numteams > 0) THEN
    RAISE EXCEPTION SQLSTATE '90001' USING MESSAGE = 'Duplicate team name';
  END IF;

  insert into teams (name, creator) values (p_name, auth.uid()) RETURNING teamid into v_teamID;
  insert into team_members (teamid, memberid, role) values (v_teamID, auth.uid(), 'owner');
  return v_teamID;
end;
$$;
```

**Comment by GaryAustin1** [comment]
RPC functions certainly give total control over access to the tables with the logic in them if you have tables blocked in RLS.   You can also set grants on their use to authenticated role, not even allowing anon access if desired.
